### DISCUSSION TITLE
Using graphql-ws client in tests

### DISCUSSION THREAD

**Discussion by Morishiri** [enisdenjo/graphql-ws#336] [Q&A] [answered]
Hello,

I try to write some tests using graphql-ws Client you provide. I based on the README examples and came up with the following:

1. Test for receiving updates

    I have a model and I want to be notified when the model changes, so I'm subscribing to it and then executing the mutation:

    ```javascript
    it('should allow to subscribe to object update', async () => {
      // const oldObject = ... (comes from fetch call in beforeEach)
      const changes = {
        name: 'newname',
      };
      return new Promise((resolve, reject) => {
        let unsubscribe;
        const onNext = ({ data, errors }) => {
          assert.strictEqual(errors, undefined);
          assert.deepStrictEqual(data, {
            objectUpdate: { object: _.merge(oldObject, changes) },
          });
          unsubscribe();
        };

        unsubscribe = userWsClient.subscribe(
          // simple helper returning correct subscription query and variables
          objectUpdate({
            parameters: {
              id: defaultProperty.id,
            },
            selection: `property { ${propertyOwnerSelection} }`,
          }),
          {
            next: onNext,
            error: reject,
            complete: resolve,
          },
        );
        // how do I know here that subscribe is already done?

        userRequest({
          // simple helper returning correct mutation query and variables
          payload: updateObjectMutation({
            input: {
              id: oldObject.id,
              ...changes,
            },
            selection: `object { ${objectSelection} }`,
          }),
        }).then(() => {
          console.log('Mutation executed!');
        });
      });
    });
    ```

    Sometimes it runs into timing issues (mutation is executed before the subscribe call finishes and I'm not receiving this update from the web socket).

    How do I know for sure that the subscribe is done and I can run the mutation?

2. Test for rejecting anonymous users

    My server is not allowing anonymous users to connect (if there is no auth then it responds with 4403):

    ```javascript
      const serverCleanup = useServer(
        {
          schema, // Adding a context property lets you add data to your GraphQL operation context
          context: async (ctx) => {
            return ctx.extra.request.context;
          },
          onConnect: async (ctx) =>
            new Promise((resolve) => {
              myAuthMiddleware(ctx.extra.request, {}, () => {
                  resolve(Boolean(ctx.extra.request.user)); // if there is user - allow, otherwise 4403
              });
            }),
        },
        wsServer,
      );
    ```

    How do I test for that? `error` in subscribe call doesn't seem to be called in that situation and I'm simply getting timeout in tests, because promise is not resolving..

Any ideas? Or maybe I should use some other testing library? What do you recommend?

**Comment by enisdenjo** [comment, marked as answer]
> How do I know for sure that the subscribe is done and I can run the mutation?

As in the subscribe was successful but the subscription is not completed yet? In that case, I recommend using [`ServerOptions.onOperation`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onoperation).

On the other hand, if you mean that the subscription really completed and no further events will be emitted - simply use the sink as showcased in the ["Client usage with Promise" recipe](https://github.com/enisdenjo/graphql-ws#promise) and wait for the Promise to resolve.

> How do I test for that? error in subscribe call doesn't seem to be called in that situation and I'm simply getting timeout in tests, because promise is not resolving..

The client is probably retrying. Consider setting the [`ClientOptions.retryAttempts`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#retryattempts) to zero (0) to have the client fail immediately.

---

Additionally, I recommend checking my test suites in [`__tests__`](https://github.com/enisdenjo/graphql-ws/tree/master/src/__tests__). Be free to ask me about parts of the tests if they're unclear - I'd be more than happy to elaborate.

**Reply by Morishiri** [reply]
Thanks for the hint with `retryAttempts`, it really worked :)

When asking about the subscription completion I asked about the case when subscribe was successful, but the subscription is not completed yet. I was hoping to get this information on client side, but having it in server's `onOperation` is good enough. Thanks.

**Reply by enisdenjo** [reply]
Awesome!

> ... I was hoping to get this information on client side

The client does know whether the subscribe is successful - when the `client.subscribe` doesn't errored out. However, it doesn't know when _exactly_ was the subscribe successful - this is only known by the server.
